### ISSUE TITLE
Why does calc_entropy_loss use log_alpha

### ISSUE THREAD

**Issue by Meruem10** [ku2482/sac-discrete.pytorch#3]
Hi,

Why do you use the log_alpha for the entropy-term:

entropy_loss = -torch.mean(
            **self.log_alpha** * (self.target_entropy - entropies).detach()
            * weights)
Furthermore, you state at the beginning, that you optimize log_alpha instead of alpha. Is that the reason why? And why you decided to do this?

Thank you in advance!

**Comment by toshikwa** [comment]
Hi, @Meruem10.

The reason why we optimize log(alpha) instead of alpha is that the numeric range is more well-behaved.
(You can see it when you try to optimize alpha.)
Actually, [Original SAC implementation](https://github.com/rail-berkeley/softlearning) did the same and I just followed it.

According to equation (17) of [SAC paper](https://arxiv.org/abs/1812.05905), optimal alpha is calculated as below.

`alpha^* = arg min_{alpha} (-alpha (target_entropy - E[entropy])).`

Because the logarithm is monotonous, this equation with log(alpha) instead of alpha still holds true.
So we can optimize log(alpha) instead of alpha. (just learning rate should be cared about.)


Anyway, thank you for asking :)

**Comment by Meruem10** [comment]
Thank you! :)

**Comment by ahmadreza9** [comment]
Hello (@ku2482) sir Watanabe, 
In calc_entropy_loss function you used self.log_alpha which is initialized by `self.log_alpha = torch.zeros(1, requires_grad=True, device=self.device)`. Now self.log_alpha is equivalent to Zero and  entropy_loss will be zero. Where you change entropy_loss and how? if you didn't, calc_entropy_loss function will always return zero. Does Adam optimizer change self.log_alpha parameter?

**Comment by toshikwa** [comment]
Hi, @ahmadreza9 .

If log_alpha is 0, loss_alpha would be 0, however, the gradient is non-zero.
For example, if x > 0 and log_alpha = 0, the gradient of (x * log_alpha) with respect to log_alpha is positive. 
(because we want to make log_alpha smaller so that (x * log_alpha) would become smaller.)

Does it make sense?

**Comment by ahmadreza9** [comment]
Thanks. So optimizer tuned entropy base on gradient(slope) of log_alpha parameter and x =(self.target_entropy - entropies).

**Comment by toshikwa** [comment]
Yes, @ahmadreza9
More specifically, optimizer tunes "entropy coefficient"(alpha), which determines the effect of entropy maximization in SAC.
